src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs

```diff
@@ -14,21 +14,52 @@ namespace Internal.Runtime.CompilerHelpers
     /// </summary>
     internal static class ArrayHelpers
     {
+        /// <summary>
+        /// Helper for array allocations via `newobj` IL instruction. Dimensions are passed in as block of integers.
+        /// The content of the dimensions block may be modified by the helper.
+        /// </summary>
+        public static unsafe Array NewObjArray(IntPtr pEEType, int nDimensions, int* pDimensions)
+        {
+            EETypePtr eeType = new EETypePtr(pEEType);
+            Debug.Assert(eeType.IsArray && !eeType.IsSzArray);
+            Debug.Assert(nDimensions > 0);
+
+            // Rank 1 arrays are handled below. We should allocate an SzArray.
+            Debug.Assert(eeType.ArrayRank > 1);
+
+            // Multidimensional arrays have two ctors, one with and one without lower bounds
+            int rank = eeType.ArrayRank;
+            Debug.Assert(rank == nDimensions || 2 * rank == nDimensions);
+
+            if (rank < nDimensions)
+            {
+                for (int i = 0; i < rank; i++)
+                {
+                    if (pDimensions[2 * i] != 0)
+                        throw new PlatformNotSupportedException(SR.PlatformNotSupported_NonZeroLowerBound);
+
+                    pDimensions[i] = pDimensions[2 * i + 1];
+                }
+            }
+
+            return Array.NewMultiDimArray(eeType, pDimensions, rank);
+        }
+
         /// <summary>
         /// Helper for array allocations via `newobj` IL instruction. Dimensions are passed in as block of integers.
         /// The content of the dimensions block may be modified by the helper.
         /// </summary>
         [UnconditionalSuppressMessage("AotAnalysis", "IL3050:RequiresDynamicCode",
             Justification = "The compiler ensures that if we have a TypeHandle of a Rank-1 MdArray, we also generated the SzArray.")]
-        public static unsafe Array NewObjArray(IntPtr pEEType, int nDimensions, int* pDimensions)
+        public static unsafe Array NewObjArrayRare(IntPtr pEEType, int nDimensions, int* pDimensions)
         {
             EETypePtr eeType = new EETypePtr(pEEType);
-            Debug.Assert(eeType.IsArray);
+            Debug.Assert(eeType.IsArray && !eeType.IsSzArray);
```

Comment:
We are handling IsSzArray case a few lines below. Also, the recursive call in the IsSzArray case handling needs to call the Rare overload.